### PR TITLE
Only rechunk for median if needed

### LINKED ISSUE #7780: High memory usage for reductions.median but not reductions.mean
I'm attempting to median stack 200 x 4k x 4k images. I've been having trouble with the memory allocation so I tried to simplify my problem by going to 2d and was surprised to see that doing a median reduction can take 50 GB of memory and crash python.

This code works as expected in ipython. Almost no memory is used and this completes in less than a minute.
```python
import dask.array as da
x = da.random.normal(10, 0.1, size=(100000, 100000), chunks=(10, 100000))
y = da.reductions.mean(x, axis=1)
%time z = y.compute()
```
`CPU times: user 6min 38s, sys: 2.36 s, total: 6min 41s
Wall time: 56.3 s`

Running the following code causes python to crash after activity monitor showing it using more than 50 GB of RAM.

```python
import dask.array as da
x = da.random.normal(10, 0.1, size=(100000, 100000), chunks=(10, 100000))
y = da.reductions.median(x, axis=1)
%time z = y.compute()
```
Screen Shot 2021-06-09 at 4 32 29 PM

Screen Shot 2021-06-09 at 4 47 17 PM

**Anything else we need to know?**:
I did also try this in regular python (as opposed to ipython) and got the same crash. The odd chunk size is an attempt to make sure each chunk has the full array necessary to compute the median in one chunk.

**Environment**:

- Dask version: 2021.05.1
- Python version: 3.8.10
- IPython version: 7.22.0
- Operating System: MacOS 11.3.1
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by jsignell**
- [x] Closes #7780
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

From original issue:
```python
import dask.array as da
x = da.random.normal(10, 0.1, size=(100000, 100000), chunks=(10, 100000))
y = da.reductions.median(x, axis=1)
%time z = y.compute()

# CPU times: user 17min 32s, sys: 2.68 s, total: 17min 35s
# Wall time: 1min 36s
```

### CODE REVIEW

File: dask/array/reductions.py [hunk 1]
```diff
@@ -1583,7 +1583,16 @@ def median(a, axis=None, keepdims=False, out=None):
...
-    a = a.rechunk({ax: -1 if ax in axis else "auto" for ax in range(a.ndim)})
+    # rechunk if reduced axes are not contained in a single chunk
+    # cannot use `any` to check since `any` is defined in this file
+    needs_rechunk = False
+    for ax in axis:
+        if a.numblocks[ax] > 1:
+            needs_rechunk = True
+            break
+
+    if needs_rechunk:
```
Comment (mrocklin) [inline dask/array/reductions.py:1594]: Maybe this?

```suggestion
    if any(a.numblocks[ax] > 1 for ax in axis):
```

File: dask/array/tests/test_reductions.py [hunk 2]
```diff
@@ -702,6 +703,24 @@ def test_median(axis, keepdims, func):
...
+def test_median_does_not_rechunk_if_whole_axis_in_one_chunk(axis, func):
+    x = np.arange(100).reshape((2, 5, 10))
+    d = da.from_array(x, chunks=(2, 1, 10))
+
+    actual = getattr(da, func)(d, axis=axis)
+    expected = getattr(np, func)(x, axis=axis)
+    assert_eq(actual, expected)
+    does_rechunk = any(
+        ["rechunk" in key_split(k) for k in actual.__dask_graph__().keys()]
+    )
```
Comment (mrocklin) [inline dask/array/tests/test_reductions.py:717]: Way dumber, but also more robust to future changes in graph structure/names

```suggestion
    does_rechunk = "rechunk" in str(dict(actual.__dask_graph__()))
```

**Comment by mrocklin** [comment]
In general this looks good to me +1

File: dask/array/reductions.py (hunk 1 above)
Comment (jsignell) [reply dask/array/reductions.py:1594]: I put a comment above, but maybe it's not clear. `any` is overridden in this file.

Comment (mrocklin) [reply dask/array/reductions.py:1594]: Ah, my apologies for not reading.  We should probably enable devs to use any.  Maybe the following?

```python
import builtins  # at the top of the file
```

```python
if builtins.any(...)
```

Comment (jsignell) [reply dask/array/reductions.py:1594]: ooh did not know about that :)

**Comment by jsignell** [comment]
Thanks for the review @mrocklin! I'll merge when green.

**Pull request merged**